linkerd/proxy/detect/src/lib.rs

```diff
@@ -62,25 +47,63 @@ impl<D, A> Accept<D, A> {
     }
 }
 
-impl<T, D, A> tower::Service<(T, BoxedIo)> for Accept<D, A>
+impl<D: Clone, S> tower::layer::Layer<S> for DetectProtocolLayer<D> {
+    type Service = DetectProtocol<D, S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        DetectProtocol {
+            detect: self.detect.clone(),
+            peek_capacity: self.peek_capacity,
+            inner,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct DetectProtocol<D, S> {
+    detect: D,
+    peek_capacity: usize,
+    inner: S,
+}
+
+pub enum DetectProtocolFuture<T, D, S>
+where
+    S: tower::Service<(D::Target, BoxedIo)>,
+    D: Detect<T>,
+{
+    Detected(S::Future),
+    Detecting {
+        inner_svc: S,
+        detect: D,
+        inner: PeekAndDetect<T, D>,
+    },
+}
+
+impl<T, D, S> tower::Service<(T, BoxedIo)> for DetectProtocol<D, S>
 where
     D: Detect<T>,
-    A: core::listen::Accept<(D::Target, BoxedIo)> + Clone,
+    S: tower::Service<(D::Target, BoxedIo)>,
+    S: Clone,
```

Comment:
nit:
```suggestion
    S: tower::Service<(D::Target, BoxedIo)> + Clone,
```

Why not continue to use the `Accept` alias though? is this not still an `Accept`?